edb/server/compiler/compiler.py

```diff
@@ -930,6 +930,14 @@ def describe_database_restore(
             and dump_server_ver.stage is not verutils.VersionStage.DEV
         )
 
+        dump_with_dunder_type = (
+            (
+                dump_server_ver is None
+                or dump_server_ver < (3, 0),
+            )
+            and dump_server_ver.stage is not verutils.VersionStage.DEV
```

Comment:
I think we need to actually start recording the catalog version in dumps to avoid this imprecise dance.